### ISSUE TITLE
Results handling

### ISSUE THREAD

**Issue by K-Phoen** [K-Phoen/rulerz#18]
Right now, the type of the results returned by `$rulerz->filter(…)` depends on the target. For instance, using a Doctrine Query Builder will return a [Collection](http://www.doctrine-project.org/api/common/2.2/class-Doctrine.Common.Collections.Collection.html) of objects hydrated by doctrine but using the Elasticsearch PHP client will return the raw Elasticsearch result as an array.

Having different result types depending on the target defeats one of the purpose of this library: the encapsulation of the storage layer.

By default, the return type for all the targets should be the same (the results with no metadata/noise).

The results could implement an interface like the following:

``` php
interface FilterResult
{
    public function getIterator();
}
```

It would for instance allow the executors to lazily return the results (using `yield` or any other method). If the number of results can easily be computed, both the `FilterResult` and the `Countable` interfaces can be implemented.

There can be cases where the user needs more control on the returned results (pagination for instance). RulerZ should provide a way to _update_ a QueryBuilder (when Doctrine is used) and return it as-is, leaving the user in charge of finishing and executing the request.
This could be done using an option in the execution context: `['result_type' => 'raw_query']`, or by exposing a new method in RulerZ.

**Comment by mnapoli** [comment]
> or by exposing a new method in RulerZ.

:+1: 

If `filter()` is kept to return all the matching results directly (as an array) that's great: it's very easy to use, it does the job it's supposed to do, it's consistent across implementations, and it keeps backward compatibility (AFAIK?). If another method is added to allow to create a filter query (the result will depend on the backend, but IMO that's to be expected because it's clear), we are then free to customize the query and run it how we want (e.g. to use pagination, etc.).

Btw I thought I would have time this week to help but I'm a little short.

**Comment by K-Phoen** [comment]
Fixed in v0.15.0

**Comment by mnapoli** [comment]
awesome!
